rpcs3/rpcs3qt/emu_settings.cpp

```diff
@@ -146,25 +152,62 @@ emu_settings::Render_Creator::Render_Creator()
 #endif
 
 #if defined(WIN32) || defined(HAVE_VULKAN)
-	// check for vulkan adapters
-	vk::context device_enum_context;
-	u32 instance_handle = device_enum_context.createInstance("RPCS3", true);
+	// Some drivers can get stuck when checking for vulkan-compatible gpus, f.ex. if they're waiting for one to get
+	// plugged in. This whole contraption is for showing an error message in case that happens, so that user has
+	// some idea about why the emulator window isn't showing up.
+	std::mutex mtx;
+	std::condition_variable cond;
+	bool thread_running = true;
+
+	QStringList compatible_gpus;
 
-	if (instance_handle > 0)
+	std::thread enumerator_thread([&]()
 	{
-		device_enum_context.makeCurrentInstance(instance_handle);
-		std::vector<vk::physical_device>& gpus = device_enum_context.enumerateDevices();
+		vk::context device_enum_context;
+		u32 instance_handle = device_enum_context.createInstance("RPCS3", true);
 
-		if (!gpus.empty())
+		if (instance_handle > 0)
 		{
-			//A device with vulkan support found. Init data
-			supportsVulkan = true;
+			device_enum_context.makeCurrentInstance(instance_handle);
+			std::vector<vk::physical_device> &gpus = device_enum_context.enumerateDevices();
 
-			for (auto& gpu : gpus)
+			if (!gpus.empty())
 			{
-				vulkanAdapters.append(qstr(gpu.get_name()));
+				// A device with vulkan support found. Init data
+				supportsVulkan = true;
+
+				for (auto &gpu : gpus)
+				{
+					compatible_gpus.append(qstr(gpu.get_name()));
+				}
 			}
 		}
+		{
+			std::scoped_lock<std::mutex> lck(mtx);
+			thread_running = false;
+		}
+		cond.notify_all();
+	});
+
+	{
+		std::unique_lock<std::mutex> lck(mtx);
+		// Loop is used to prevent sprudious wakeups from triggering false positives
+		while (thread_running && cond.wait_for(lck, std::chrono::seconds(10)) != std::cv_status::timeout);
+	}
+
+	if (thread_running)
+	{
+		LOG_FATAL(GENERAL, "Vulkan device enumeration timed out");
+		QMessageBox::critical(nullptr, tr("Vulkan check timeout"),
+			tr("Querying for Vulkan-compatible devices is taking too long. This is usually caused by malfunctioning "
+			"graphics drivers, reinstalling them could fix the issue."),
+			QMessageBox::Abort, QMessageBox::Abort);
+		std::exit(1); // The thread hasn't been detached, so it will trigger std::terminate anyways
```

Comment:
Why exit? OpenGL is still usable for affected users. I feel it would be a bit heavy-handed to forcefully exit without letting the user decide.

Reply:
I originally considered that as well, but I was worried that there would be some issues when running the emulator with partly working vulkan and I figured that the user should sort out their drivers anyways. But I guess there's no harm in leaving the thread running in background and letting users continue, hopefully the delay and message are annoying enough that they'll look into how to fix it.

Reply:
Consider using a named_thread and lowering the priority before making the call. That was at least in the worst case it runs silently in the background. We could also just ask the user to choose whether to terminate or proceed with OpenGL only.

Reply:
What about the local variables though? I'm worried that if the thread somehow unfreezes itself, it'll start accessing variables that fell out of scope after the function returned.

Reply:
making all of them static thread_local solves tgis issue.

Reply:
the thread_local is only if this function can be accessed from more than one thread.